### ISSUE TITLE
Transactions?

### ISSUE THREAD

**Issue by tych0** [rqlite/rqlite#266]
Hi,

I'm interested in some sort of transaction functionality beyond what rqlite currently provides today. In particular, I'd like to be able to get the results of a query in the middle of a transaction, as opposed to just "process this set of queries", the primitive that rqlite provides today.

We could do transactions as the leader, applying them to the sqlite database and accumulating the queries to apply until the transaction was committed, but I don't think that would actually guarantee correctness (someone else who isn't in the transaction could take e.g. a primary key ID that was allocated in the transaction, since we're not committing things to the raft logs as they're added to the tx).

Do you have any ideas on how we'd go about doing this, or if it's even possible?

Thanks

**Comment by tych0** [comment]
Actually, after a careful reading of:

https://www.sqlite.org/lang_transaction.html
https://www.sqlite.org/lockingv3.html

I think that this is possible. It would look something like this:

1. The leader does a BEGIN and returns a tx id.
2. That transaction accumulates queries (by using golang sql's tx.Execute/tx.Query methods to return results)
3. when done, the leader acquires a lock, commits the transaction to the sqlite locally, and then also commits the transaction to the raft logs. Once the transaction is committed to the raft logs, the lock is released.

I think this is safe because active transactions will never overlap with each other, so the potential race here is that things are committed in the wrong order. Since we only ever allow the leader to do transactions, and the leader synchronizes transactions in the order they're committed, things should be safe.

I'll see if I can come with an implementation.

**Comment by tych0** [comment]
FWIW, there's an implementation of this here: https://github.com/tych0/rqlite/tree/transaction

The queries don't quite work, but I think this is likely just some mechanics about converting values from query results to actual usable values. I'll keep fiddling with that, but the transaction approach is there (and it is also quite small, all things considered).

**Comment by otoolep** [comment]
Interesting! I think it's safe, but there might be one problem.

I do agree that if the leader only commits to the Raft log after successfully committing the transaction, it's pretty safe. Since, on start-up, the SQLite database is completely rebuilt from the Raft log any commits in the database, but not in the log, will not be applied to the newly-built database under the node.

However, there is a concept of snapshotting the SQLite database, as a part of compacting the log (and this state is shipped to other nodes that join). I'm not sure what guarantees there are around when that snapshot is taken. If it was taken while the commits were in the database, but before the log was applied to the database, it would be trouble.

Perhaps the Hashicorp Raft documentation can rule out this possibility? If so, I think your idea is really interesting, and could be quite useful.

**Comment by tych0** [comment]
Thanks, I'll check the docs out. I think we should be locking while taking a snapshot anyway, but it seems like there's some time where the snapshot is in the leader, but not all the way through the raft machinery, and we should probably take care to avoid a race there.

**Comment by tych0** [comment]
Actually, I think that locking during Snapshot() should be enough: the race we're concerned about is one where the db has something that the raft log doesn't. Assuming we synchronize on the same lock when committing transactions and taking snapshots, I think this should be okay. (In fact, we have to do this so that we don't corrupt the SQLite databse.)

One issue is what to do here: https://github.com/tych0/rqlite/commit/a7c0daab7b5357336f859c004a358baff7d8f2b6#diff-1f8872a5d54402aa220c723deb16293cR176

That is a very real error where somehow we fail to commit the logs, but we succeeded to commit the transaction. I suppose we could snapshot before every commit and rollback, but that seems quite expensive. The other option is just to force a rebuild from the raft log somehow.

**Comment by tych0** [comment]
FWIW, I just posted an implementation in the PR above.

**Comment by otoolep** [comment]
Since you submitted a second PR, I just want to be sure I understand how you propose this should work.

If a client wants to get results of a query in the middle of a transaction (which is the example reason you give and a valid use case), the client must explicitly ask to start a transaction. All good, the node talks to the SQLite database directly without going through the log (which it never does today), and the client gets an ID back.

Now, while is this transaction is active, the client performs the _execs_ and _queries_, and these queries are performed against the SQLite database directly -- not through the Raft log. Once the client is happy, it calls _commit_, the node takes the list of commands that were executed, wraps them up in a single log message, and attempts to send the log message through Raft. If it fails, wipe the database directly and rebuild.

Some things that occur to me: 

-- for the transaction situation, the rqlite state machine code would need to change such that it just commits those entries to the log, but doesn't apply the log to the database, on the leader (since the leader will have already modified its local SQLite database). But a follower would still need to apply the log messages to its local database since it would not be involved directly with the transaction handling. This makes me nervous -- to have a node handle messages differently depending on whether it is a leader or a follower. And the problem with a node checking if it's a leader is that leadership status can change at anytime. For a node to be absolutely sure it is the leader, it requires a round trip through the Raft consensus mechanism -- this means another message in the log. See _strong_ read consistency to see this in action.

-- what would happen if the leader handling the transaction was deposed? Due to a network fault? The call to `Apply()` would fail, which might seem OK. But what would happen if, during the transaction processing, the leader was deposed, was not the leader for a while, and then re-elected so the call to `Apply()` was successful? But the SQLite database under the leader may no longer be what is was before, since the other node that was leader for a short while may have made changes. Does that make sense?

These are just cases that I think are difficult to address. I think the change you're proposing is ambitious, I think the edge cases are hard to reason and be sure they've been solved.

Let me know what you think.

**Comment by tych0** [comment]
On Mon, Feb 13, 2017 at 09:35:51PM -0800, otoolep wrote:
> Since you submitted a second PR, I just want to be sure I understand how you propose this should work.
> 
> If a client wants to get results of a query in the middle of a transaction (which is the example reason you give and a valid use case), the client must explicitly ask to start a transaction. All good, the node talks to the SQLite database directly without going through the log (which it never does today), and the client gets an ID back.
> 
> Now, while is this transaction is active, the client performs the _execs_ and _queries_, and these queries are performed against the SQLite database directly -- not through the Raft log. Once the client is happy, it calls _commit_, the node takes the list of commands that were executed, wraps them up in a single log message, and attempts to send the log message through Raft. If it fails, wipe the database directly and rebuild.
> 
> Some things that occur to me: 
> 
> -- for the transaction situation, the rqlite state machine code would need to change such that it just commits those entries to the log, but doesn't apply the log to the database, on the leader (since the leader will have already modified its local SQLite database). But a follower would still need to apply the log messages to its local database since it would not be involved directly with the transaction handling. This makes me nervous -- to have a node handle messages differently depending on whether it is a leader or a follower. And the problem with a node checking if it's a leader is that leadership status can change at anytime. For a node to be absolutely sure it is the leader, it requires a round trip through the Raft consensus mechanism -- this means another message in the log. See _strong_ read consistency to see this in action.

I don't understand what you mean here: why does the follower need to
be aware of the transaction at all, until it is committed via raft
log?

> -- what would happen if the leader handling the transaction was deposed? Due to a network fault? The call to `Apply()` would fail, which might seem OK. But what would happen if, during the transaction processing, the leader was deposed, was not the leader for a while, and then re-elected so the call to `Apply()` was successful? But the SQLite database under the leader may no longer be what is was before, since the other node that was leader for a short while may have made changes. Does that make sense?

Great question, and I hope this is addressed by this code:
https://github.com/rqlite/rqlite/pull/279/commits/31784700a7842886e09ff1fb69045aad5b50fd63#diff-1f8872a5d54402aa220c723deb16293cR155

Basically, once a transaction has failed a raft Apply, you can't do
anything but roll it back, exactly because of this problem.

**Comment by tych0** [comment]
Oh, and I should confirm: your summary is correct, that's how the implementation works right now.

**Comment by otoolep** [comment]
>>I don't understand what you mean here: why does the follower need to be aware of the transaction at all, until it is committed via raft log?

Let me take a step back. Today, when any node has `Apply()` called on its `fsm`by the Raft consensus mechanism, the code unpacks the message in the Raft log entry -- one or more SQLite statements -- and applies them to the underlying database. Each node acts exactly the same way. Now, with this new design, it seems to me that the leader will have to act differently than the followers -- it can't apply the SQLite commands to the database contained in the Raft log entry, during the execution of `Apply()`, *because it will have already applied them*. Only followers would be safe applying the SQLite commands contained in the log entry.

Assuming I am correct, this introduces a few issues. Fundamentally for a node to check that it is the leader (so that `Apply()` executes differently) it must perform a distributed consensus operation -- it much ask each other node if it's the leader in a very specific manner. It's not enough to just check some local state -- not if you want a 100% answer (which we do for any activity that may result in changes to the underlying SQLite database). All this is not trivial. Does this make sense? Or am I missing something?

>>Great question, and I hope this is addressed by this code:

Unfortunately I don't think it does. I might be wrong, but it would need rigorous testing to be sure. The issue is that there is nothing preventing node A being the leader, node B suddenly being elected leader (network fault disconnects A from cluster), node B being deposed (network fault disconnects B, but A is bakc), and node A being elected leader again, any number of times, between these two lines of code:

`err := t.tx.Commit()`

and

`_, err = t.store.Execute(t.stmts, false, true)`

Assuming that the node is leader during both of these lines of code, and that the Raft term has not changed, is an invalid assumption (as far as I know). Assuming I am correct, I have no idea how this change would interact with the system going through an election cycle above.

**Comment by tych0** [comment]
On Tue, Feb 14, 2017 at 07:33:58AM -0800, otoolep wrote:
> >>I don't understand what you mean here: why does the follower need to be aware of the transaction at all, until it is committed via raft log?
> 
> Let me take a step back. Today, when any node has `Apply()` called on its `fsm`by the Raft consensus mechanism, the code unpacks the message in the Raft log entry -- one or more SQLite statements -- and applies them to the underlying database. Each node acts exactly the same way. Now, with this new design, it seems to me that the leader will have to act differently than the followers -- it can't apply the SQLite commands to the database contained in the Raft log entry, during the execution of `Apply()`, *because it will have already applied them*. Only followers would be safe applying the SQLite commands contained in the log entry.

Ah ha, I was missing this point that the leader has its Apply() called
too; makes sense.

> Assuming I am correct, this introduces a few issues. Fundamentally for a node to check that it is the leader (so that `Apply()` executes differently) it must perform a distributed consensus operation -- it much ask each other node if it's the leader in a very specific manner. It's not enough to just check some local state -- not if you want a 100% answer (which we do for any activity that may result in changes to the underlying SQLite database). All this is not trivial. Does this make sense? Or am I missing something?

It does make sense. Suppose we change the log messages to include a
uuid for the transaction; then the store can look at its active
transaction and not apply it if the uuids match?

> >>Great question, and I hope this is addressed by this code:
> 
> Unfortunately I don't think it does. I might be wrong, but it would need rigorous testing to be sure. The issue is that there is nothing preventing node A being the leader, node B suddenly being elected leader (network fault disconnects A from cluster), node B being deposed (network fault disconnects B, but A is bakc), and node A being elected leader again, any number of times, between these two lines of code:
> 
> `err := t.tx.Commit()`
> 
> and
> 
> `_, err = t.store.Execute(t.stmts, false, true)`

I don't think the leader changing is a problem, but you do raise a
good point: if A is the leader and starts a transaction, then B gets
elected and starts a transaction, then A gets re-elected and commits,
followed by B's re-election and commit, B's commit succeed, but really
fail silently if they both acquired the same rowid for something and
used it in a subsequent insert.

I think the right thing to do here is simply to invalidate any active
transactions when a new leader is elected. Does that make sense?

**Comment by otoolep** [comment]
@tych0 -- some of the ideas you raise might work. But I think you can see now that what you are suggesting is not trivial, and has lots of corner cases that need extensive testing. Without this kind of rigorous testing, I would be very hesitant to merge the proposed transaction support into the master. 

In its current form rqlite provides very strong correctness guarantees due to the way it uses the Raft consensus module. But a system that attempts to provide transactions in the form you suggest would not do so. What we really need is something more like standard 2-phase commit.

**Comment by otoolep** [comment]
>>then the store can look at its active transaction and not apply it if the uuids match?

It might, but it's starting to get complicated.

>>I think the right thing to do here is simply to invalidate any active
transactions when a new leader is elected. Does that make sense?

Assuming you can even hook into when this happens, in a way that guarantees you won't miss the election, and it's not racy.

**Comment by tych0** [comment]
On Tue, Feb 14, 2017 at 08:05:03AM -0800, otoolep wrote:
> >>then the store can look at its active transaction and not apply it if the uuids match?
> 
> It might, but it's starting to get complicated.

Naturally :)

> >>I think the right thing to do here is simply to invalidate any active
> transactions when a new leader is elected. Does that make sense?
> 
> Assuming you can even hook into when this happens, in a way that guarantees you won't miss the election, and it's not racy.

I think you can:
https://godoc.org/github.com/hashicorp/raft#LeaderObservation

**Comment by tych0** [comment]
On Tue, Feb 14, 2017 at 08:03:10AM -0800, otoolep wrote:
> @tych0 -- some of the ideas you raise might work. But I think you can see now that what you are suggesting is not trivial, and has lots of corner cases that need extensive testing. Without this kind of rigorous testing, I would be very hesitant to merge the proposed transaction support into the master. 

Believe me, I could see it before too :). The point of this thread is
not to get things merged in to master as is, but to arrive at a
consensus (!) that I can implement.

> In its current form rqlite provides very strong correctness guarantees due to the way it uses the Raft consensus module. But a system that attempts to provide transactions in the form you suggest would not do so. What we really need is something more like standard 2-phase commit.

Ok. Do you have a suggestion for what that would look like?

**Comment by otoolep** [comment]
>>I think you can: https://godoc.org/github.com/hashicorp/raft#LeaderObservation

Since that's not part of the Raft core, but a convenience feature that Hashicorp built, I'm not sure what the guarantees are.

>>Ok. Do you have a suggestion for what that would look like?

I know, at a high-level, what two-phase commit is -- Raft is sort of a 2PC system AFAIK, but I have never coded one, so you're running up against the limitation of my knowledge now.

I have another idea, which I have never had the time to explore. Perhaps it would work? Let me explain, perhaps you can try coding it.

A distributed transaction is started as follow -- the client requests to start a transaction, and the SQL statement `BEGIN` is sent through the Raft log in the normal way. Each node will get this statement per the usual methods, and open a transaction on each underlying database. Now each SQLite database is in the same state. The client sends further _execs_ and _queries_ as usual, all of which are taking place within the transaction. All these commands go through the log.

Finally the client decides to commit, or rollback. Again, the system simply sends `COMMIT` or `ROLLBACK` through the log, and each node processes the command as usual. Each database ends up in the same state. If `COMMIT` fails, well, the transaction is aborted on each node as usual. The commands remain in the log, but it's not a problem. At the end of the day the database under each node is in the same state. In other words the system operates the same way as ever. 

The README states that commands such as `COMMIT` are not supported by rqlite, but this is only because I never tested the above scenario and didn't want to make any promises. But perhaps it would all work nicely.

There are a few things to consider.

-- a node crashes while holding a transaction. In this case the transaction will be aborted automatically. However, when the node restarts the database is wiped and rebuilt from what's in the logs. It will recover to the state it was, and pick up any commands that occurred while it was down. 

-- the entire cluster crashes during a transaction. In that case, the cluster will come back up in state that finds itself in the middle of a transaction. This could be rather strange, since one would imagine that a cluster restart should clear out any transaction. There might be case here for rolling back any transaction that system finds itself in, due to a replay of a log on restart. In other words, after the log is applied after starting up, issue an automatic `ROLLBACK` by default. If there is no transaction in progress, it's not a big deak.

-- clients, other than the client that is using the transaction, should be allowed to read the database while a transaction is in place. However, those other clients should not be permitted to write to the database while a transaction is outstanding. So to prevent this, a transaction ID (or something) should be handed out to the client that starts the transaction, and that client should pass back the transaction ID. It might be possible to use the transaction ID returned by the command `BEGIN`. 

Does this make sense? Do you want to see if sending `BEGIN`, `COMMIT` etc through the Raft system would actually work?

**Comment by tych0** [comment]
On Feb 14, 2017 3:37 PM, "otoolep" <notifications@github.com> wrote:

I think you can: https://godoc.org/github.com/hashicorp/raft#
LeaderObservation

Since that's not part of the Raft core, but a convenience feature that
Hashicorp built, I'm not sure what the guarantees are.

Ok. Do you have a suggestion for what that would look like?

I know, at a high-level, what two-phase commit is -- Raft is sort of a 2PC
system AFAIK, but I have never coded one, so you're running up against the
limitation of my knowledge now.

I have another idea, which I have never had the time to explore. Perhaps it
would work? Let me explain, perhaps you can try coding it.

A distributed transaction is started as follow -- the client requests to
start a transaction, and the SQL statement BEGIN is sent through the Raft
log in the normal way. Each node will get this statement per the usual
methods, and open a transaction on each underlying database. Now each
SQLite database is in the same state. The client sends further *execs* and
*queries* as usual, all of which are taking place within the transaction.
All these commands go through the log.

Finally the client decides to commit, or rollback. Again, the system simply
sends COMMIT or ROLLBACK through the log, and each node processes the
command as usual. Each database ends up in the same state. If COMMIT fails,
well, the transaction is aborted on each node as usual. The commands remain
in the log, but it's not a problem. At the end of the day the database
under each node is in the same state. In other words the system operates
the same way as ever.

The README states that commands such as COMMIT are not supported by rqlite,
but this is only because I never tested the above scenario and didn't want
to make any promises. But perhaps it would all work nicely.

There are a few things to consider.

-- a node crashes while holding a transaction. In this case the transaction
will be aborted automatically. However, when the node restarts the database
is wiped and rebuilt from what's in the logs. It will recover to the state
it was, and pick up any commands that occurred while it was down.

What will automatically abort the transaction in this case? This seems
quite hard to detect.


-- the entire cluster crashes during a transaction. In that case, the
cluster will come back up in state that finds itself in the middle of a
transaction. This could be rather strange, since one would imagine that a
cluster restart should clear out any transaction. There might be case here
for rolling back any transaction that system finds itself in, due to a
replay of a log on restart. In other words, after the log is applied after
starting up, issue an automatic ROLLBACK by default. If there is no
transaction in progress, it's not a big deak.

-- clients, other than the client that is using the transaction, should be
allowed to read the database while a transaction is in place. However,
those other clients should not be permitted to write to the database while
a transaction is outstanding. So to prevent this, a transaction ID (or
something) should be handed out to the client that starts the transaction,
and that client should pass back the transaction ID. It might be possible
to use the transaction ID returned by the command BEGIN.

Does this make sense? Do you want to see if sending BEGIN, COMMIT etc
through the Raft system would actually work?


Sounds reasonable. Another problem is parallel transactions, which this
won't allow for. But perhaps we can work some magic around that.


—

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/rqlite/rqlite/issues/266#issuecomment-279859037>, or mute
the thread

**Comment by tych0** [comment]
Euh, looks like github butchered my email reply:

> -- a node crashes while holding a transaction. In this case the transaction
> will be aborted automatically. However, when the node restarts the database
> is wiped and rebuilt from what's in the logs. It will recover to the state
> it was, and pick up any commands that occurred while it was down.

What will automatically abort the transaction in this case? This seems
quite hard to detect.


> Does this make sense? Do you want to see if sending BEGIN, COMMIT etc
> through the Raft system would actually work?


Sounds reasonable. Another problem is parallel transactions, which this
won't allow for. But perhaps we can work some magic around that.

**Comment by otoolep** [comment]
>What will automatically abort the transaction in this case? This seems
quite hard to detect.

Since the connection to SQLite will be dropped, I presume SQLite itself does it -- or handles it.

**Comment by tych0** [comment]
Hmm. But sqlite isn't a daemon or anything; if for example golang segfaults, sqlite isn't running any more.

**Comment by otoolep** [comment]
>Hmm. But sqlite isn't a daemon or anything; if for example golang segfaults, sqlite isn't running any 
more.

Obviously I know that. :-) What I mean is that any lingering transaction might be cleaned up the next time SQlite is accessed, or something. I have not pushed on this myself. Perhaps you will know better.

**Comment by tych0** [comment]
Ok. I don't see how sqlite would know that a transaction was lingering (and then how it would communicate it back to us so that we can ROLLBACK in the raft log), but I'll see what I can dig up.

**Comment by otoolep** [comment]
I don't know for a fact either, and I have not tested it. But I'd be shocked if the SQLite system had not been implemented to handle a client crashing in the middle of a transaction. Maybe that transaction is left dangling forever, but with zero practical impact on the system. When the node restarted, it would be restarting a new transaction.

**Comment by tych0** [comment]
Sure sqlite may have support for it, but we're not really dealing with sqlite, we're dealing with our own log of things. The problem here is that you can't nest transactions:

```
$ sqlite3 :memory:
SQLite version 3.11.0 2016-02-15 17:29:24
Enter ".help" for usage hints.
sqlite> BEGIN;
sqlite> BEGIN;
Error: cannot start a transaction within a transaction
sqlite> 
```

So consider:

1. Node issues a BEGIN
2. Node crashes
3. Node restarts, and re-builds the sqlite3 database from the raft log.
4. Nobody can BEGIN, because nothing ROLLBACK'd the previous transaction.

All the sqlite3 intelligence about crashing doesn't help us here, because our basic primitive is not sqlite3, but statements in the raft log.

I suppose we could look at the old sqlite DB from the last boot and see if it had a partially written transaction, but there's still no way to detect if that partial transaction was as a result of the crash, or if it crashed during the Apply() of a COMMIT from some raft neighbor.

**Comment by otoolep** [comment]
Sure. If you look back on what I wrote yesterday, I called this out:

"-- the entire cluster crashes during a transaction. In that case, the cluster will come back up in state that finds itself in the middle of a transaction. This could be rather strange, since one would imagine that a cluster restart should clear out any transaction. "

Perhaps I am missing your point.

**Comment by otoolep** [comment]
If there is a way to deal with this, it would be great. But it may be the user would have no choice but to issue an explicit ROLLBACK after a restart. Perhaps there is some way to handle it automatically.

**Comment by otoolep** [comment]
If there is a way to deal with this, it would be great. But it may be the user would have no choice but to issue an explicit ROLLBACK after a restart. Perhaps there is some way to handle it automatically.

**Comment by otoolep** [comment]
Closed the issue in error.

**Comment by tych0** [comment]
Right, I guess I think it's possible if just one node crashes, not only the entire cluster. An explicit rollback after a single node failure would be unfortunate.

**Comment by tych0** [comment]
I suppose we could ROLLBACK after leader election, and only allow the leader to issue BEGIN and COMMIT/ROLLBACK statements.

**Comment by otoolep** [comment]
Yeah, this is not looking as easy as I first thought. I knew there was a reason the system didn't officially support BEGIN and friends. :-)

**Comment by tych0** [comment]
A third idea I had thought about was to use actual sqlite dumps as raft log entries, instead of the queries themselves. It would be unfortunate to have to transmit the dump every time, though, and when I poked around there's no way to serialize transactions in a binary format that sqlite can apply. But anyway, it's a thought :)

**Comment by MrHacky** [comment]
https://github.com/rqlite/rqlite/issues/266

as i am also in need of this functionality and was reading through this issue and thinking about the problems with it, i would first like to clear up a misunderstanding about sqlite.

Specifically, if there is a crash after sqlite BEGIN but before COMMIT, the next time you open the database it will be as if none of the statements after BEGIN ever happened. sqlite has its own log/journal and only a COMMIT will make a sqlite transaction permanent.

Also i thought of a possible trick to handle some of the problems mentioned:
- The leader does a BEGIN and returns a tx id.
- That transaction accumulates queries (and executes them to return results)
- when done, execute a ROLLBACK abandoning the sqlite transaction, and use the list of accumulated queries as a normal raft log entry

**Comment by otoolep** [comment]
Thanks, for your input, @MrHacky 

>Specifically, if there is a crash after sqlite BEGIN but before COMMIT, the next time you open the database it will be as if none of the statements after BEGIN ever happened. sqlite has its own log/journal and only a COMMIT will make a sqlite transaction permanent.

I think everyone understood that. I certainly did

Your suggestion is interesting however. It would require the node holding the transaction open to refuse all other statements, that are not part of the transaction. That's not a blocker. The real issue is when does the node give up waiting for the client to say it's done? I guess the client could specify a timeout, and if a request is not received within that time, abort it.

I am also thinking about adding support for sending queries over TCP (as opposed to HTTP). This would allow the node to abort the operation is the TCP connection was closed.

**Comment by Esowteric** [comment]
Thanks, this is a sticking point for me, too. I'm running a node.js app at the moment using a single instance of better-sqlite3, and at some point I'm pretty sure I'll need a distributed setup.

I was hoping that rqlite might be the way to go, but I need to be able to manually BEGIN a transaction (unless I'm already in a transaction) and call a function (keeping others locked out, waiting) that runs an insert statement and returns an ID. Then I do a bit more processing using that ID, and either COMMIT if everything has gone okay and I'm not already in a transaction, or ROLLBACK if I'm not already in a transaction.

This works fine with better-sqlite3 and it might work if rqlite operated statement-by-statement, including BEGIN and COMMIT or ROLLBACK, but not as it stands.  

TCP support sounds like a good idea, too.

Anyhow, thanks again: Will keep checking on progress and looking around for alternatives.

~ Eric T.

**Comment by otoolep** [comment]
@Esowteric -- to be clear, you can issue the `BEGIN` and `COMMIT` commands yourself if you wish. The point is that the behavior of rqlite under those circumstances is currently undefined. You risk leaving the system in a locked state if any problems (such as a crash) occur between `BEGIN` and `COMMIT` (or `ROLLBACK`).

**Comment by otoolep** [comment]
Initiated a discussion of this issue here: https://groups.google.com/forum/#!topic/raft-dev/BaS5Z2NkDxA

**Comment by Esowteric** [comment]
Thanks a lot for the reply and the discussion thread, @otoolep.

**Comment by freeekanayaka** [comment]
@otoolep @Esowteric you might want to give a look at [dqlite](https://github.com/CanonicalLtd/dqlite)

**Comment by otoolep** [comment]
Thanks -- I came across dqlite recently, interesting work. However it would require me to re-write the application layer since dqlite is a library that must be linked with other code (not a huge deal of course, it would just take time). I would also need a significant patch to the SQLite C source code, which would then mean rqlite could no longer claim it was built on production-grade SQLite. rqlite users couldn't have the same assurance about the database engine that they do now (though if the patch gets merged upstream it's a different story).

Let me know if I am missing something.

**Comment by freeekanayaka** [comment]
That's all correct. The patch on original SQLite is very light and it really won't compromise the quality of SQLite, although of course it's something that might make people not so happy. I'll be proposing upstream to merge the patch in the coming weeks or months, when I'll also complete some rough edges of dqlite (it currently requires quite a bit of plumbing for consumers, so I'll added some abstractions on top of the low-level Driver to make that easier).

**Comment by otoolep** [comment]
The design I am proposing for transactions is as follows.

- a new endpoint for creating a new transaction. A Client makes a HTTP `POST` request to `/db/transaction` to being this transaction. This request will eventually get down to the SQLite layer as a `BEGIN` statement.
- rqlite returns an opaque transaction ID.
- subsequent SQL write statements that are to take place within this transaction must include this transaction ID in the request, as the query param `tx_id`.
- if rqlite receives any write statement that does not include this transaction ID, rqlite will return an error (with one possible exception -- see later). The error returned to the client is basically equivalent to the "database locked" response that a SQLite client can receive today.
- the client can commit the transaction, or roll it back, by making HTTP requests to `/db/transaction/<tx_id>`. A `PUT` to this ID will commit it, a `DELETE` will roll it back.
- savepoint functionality may also be implemented, via the endpoint `db/transaction/<tx_id>/savepoint`

Failure scenarios are handled as follows:

- leader node, which is handling the transaction, crashes during transaction processing. In this case a new election will take place (as normal). The client must redirect its processing to the new leader, and it can continue where it left off. The node that crashed will come back up and rebuild its state from the Raft log as usual. It will apply any changes it missed while down, again as usual.
- leader node loses network connectivity to the cluster, due to a network partition, during transaction processing. A leader election will take place as described above, and the client must handle it accordingly.
- entire cluster goes down. When it's restarted the Raft log will be replayed and the transaction will be restarted. This means that restarting the cluster will not clear a transaction, which may be odd. However I see no way to avoid this outcome.
- the client crashes during transaction processing, and never commits or rolls back the transaction. This can be handled by another client at anytime by explicitly rolling back the transaction, if that client is sure this it the right thing to do (not necessarily trivial).
- rqlite will support associating an optional timeout with a transaction. Everytime rqlite receives a SQL statement while a transaction is active, and if the this SQL statement does not include the correct transaction ID, it will check the time the last SQL statement associated with the transaction was received. If this time is more than "timeout" in the past, the rqlite system will rollback the transaction and then execute the latest SQL statement it just received.
-- behavior or rqlite, due to a client sending `BEGIN`, `COMMIT`, `ROLLBACK` etc, will continue to be undefined.

This is the current design proposal. A key test of whether this will work will be if something that requires full transaction support can be supported by rqlite. For example, can the rqlite Python driver be made fully functional, such that something like Django can run on rqlite?

Finally, the current use of the word "transaction" in rqlite docs and code will be replaced with the word "atomic" -- with one exception. The data API will silently support `transaction` as a query param, the purposes of backward compatibility. This query param will be officially deprecated however.

**Comment by otoolep** [comment]
It may be more standard to use the HTTP header Session-Id` to return the connection ID, and support this as a way to return the connection ID in a request.

**Comment by otoolep** [comment]
The latest design for this remains a new endpoint `/db/connection` -- this work has started in https://github.com/rqlite/rqlite/pull/466

Isolated connections should make supporting transactions much easier. In fact, it may allow two modes of transactions:

-- one where every command on the new transaction, on a given connection, go through Raft. This includes `BEGIN`, `COMMIT` and `ROLLBACK` (if executed). Call this _strong_ transaction support. This would have the advantage that transaction state would carry over to every leader, if the current leader failed.
-- one where the transaction is just opened on the local SQLite database directly, without going through the Raft log. Only if the transaction was committed would the entire set of queries be committed. This would be much faster, but the issue here is that the SQLite database could be left in a LOCKED state after a leader election. If an election occurred during a transaction, this now non-leader node might reject changes from the new leader because its database was LOCKED, whereas all other nodes were not. This might be handled by the the first node listening for leader changes and immediately rolling back any transaction, but this is racy.

**Comment by otoolep** [comment]
Interesting...this commercial database has a similar transaction mechanism to one proposed above.

https://docs.marklogic.com/guide/rest-dev/transactions

**Comment by otoolep** [comment]
https://github.com/rqlite/rqlite/pull/493

**Comment by otoolep** [comment]
Implemented.
